src/Google.Cloud.Metadata.V1/MetadataClient.cs

```diff
@@ -0,0 +1,571 @@
+        /// </summary>
+        /// <remarks>
+        /// The project metadata contains information about the project, such as its ID, and custom metadata and it can be
+        /// accessed without authentication from an application running on an instance of that project. Check <see cref="IsServerAvailable"/> or
+        /// <see cref="IsServerAvailableAsync(CancellationToken)"/> to determine if the metadata can be accessed.
+        /// </remarks>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>A task containing the <see cref="InstanceMetadata"/> representing the metadata.</returns>
+        public virtual Task<ProjectMetadata> GetProjectMetadataAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Gets the value indicating whether the metadata server or an emulator for the server is available synchronously.
+        /// </summary>
+        /// <remarks>
+        /// Normally, the metadata server can only be accessed from an application running on a Google Compute Engine VM instance or
+        /// Google App Engine Flexible Environment, but if an metadata server emulator is available, this will also return true.
+        /// </remarks>
+        /// <returns>True if the normal metadata server or an emulator is available; false otherwise.</returns>
+        public virtual bool IsServerAvailable()
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Gets the value indicating whether the metadata server or an emulator for the server is available asynchronously.
+        /// </summary>
+        /// <remarks>
+        /// Normally, the metadata server can only be accessed from an application running on a Google Compute Engine VM instance or
+        /// Google App Engine Flexible Environment, but if an metadata server emulator is available, this will also return true.
+        /// </remarks>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>A task containing true if the normal metadata server or an emulator is available; false otherwise.</returns>
+        public virtual Task<bool> IsServerAvailableAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            throw new NotImplementedException();
+        }
+
+        // TODO: The http requests to modify the metadata require a fingerprint of the current metadata state to make sure you have the latest info
+        //       when making an update. This design does not allow for the same type of guarantee. Should we instead store the fingerprint on the
+        //       metadata classes internally and move these modification methods to InstanceMetadata and ProjectMetadata?
```

Comment:
That was my original thought when writing this comment, but now I'm not so sure. You can't get the fingerprint from the metadata server. You need to get it from the Compute API. So to get the instance/project metadata and ensure that the fingerprint is accurate without worrying about race conditions, we'd need to get all metadata from the Compute API and completely bypass the metadata server, even if the caller only wants to read the metadata, because we have no idea if they might try to set something later. So I don't think this can work.

I still think the best way would be to have a callback approach where an immutable dictionary is specified and the callback returns changes descriptions or an updated dictionary.